src/handlers.jl

```diff
@@ -229,6 +229,14 @@ function shutdown_request(socket, msg)
     # stop heartbeat thread
     stop_heartbeat(heartbeat[], heartbeat_context[])
 
+    # Shutdown the `requests` socket handler before sending any messages. This
+    # is necessary because otherwise the event loop will be calling
+    # `recv_ipython()` and holding a lock on `requests`, which will cause a
+    # deadlock when we try to send a message to it from the `control` socket
+    # handler.
+    global _shutting_down[] = true
+    @async Base.throwto(requests_task[], InterruptException())
```

Comment:
I think a contributing factor to this bug is that the "shutdown_reply" being sent on the SHELL/requests channel instead of the CONTROL channel. My reading of the client messaging spec is that all replies should be sent on the channel they were received on, so I believe that

```suggestion
    @async Base.throwto(socket, InterruptException())
```

or 
```suggestion
    @async Base.throwto(control[], InterruptException())
```
would be more correct (I've never seen a `shutdown_request` sent on the SHELL channel, and that using the SHELL channel for `shutdown_request`s is deprecated on 5.4 of the client messaging API) and likely avoid a deadlock from the SHELL/requests channel being locked.